src/node_shadow_realm.cc

```diff
@@ -1,15 +1,109 @@
 #include "node_shadow_realm.h"
+#include "env-inl.h"
 
 namespace node {
 namespace shadow_realm {
 using v8::Context;
+using v8::EscapableHandleScope;
+using v8::HandleScope;
 using v8::Local;
 using v8::MaybeLocal;
+using v8::Value;
+
+// static
+ShadowRealm* ShadowRealm::New(Environment* env) {
+  ShadowRealm* realm = new ShadowRealm(env);
+  env->AssignToContext(realm->context(), realm, ContextInfo(""));
+
+  if (realm->RunBootstrapping().IsEmpty()) {
+    delete realm;
+    return nullptr;
+  }
+  return realm;
+}
 
 // static
 MaybeLocal<Context> HostCreateShadowRealmContextCallback(
     Local<Context> initiator_context) {
-  return Context::New(initiator_context->GetIsolate());
+  Environment* env = Environment::GetCurrent(initiator_context);
+  ShadowRealm* realm = ShadowRealm::New(env);
+  if (realm != nullptr) {
+    return realm->context();
+  }
+  return MaybeLocal<Context>();
+}
+
+// static
+void ShadowRealm::WeakCallback(const v8::WeakCallbackInfo<ShadowRealm>& data) {
+  ShadowRealm* realm = data.GetParameter();
+  delete realm;
+}
+
+ShadowRealm::ShadowRealm(Environment* env)
+    : Realm(env, NewContext(env->isolate()), kShadowRealm) {
+  context_.SetWeak(this, WeakCallback, v8::WeakCallbackType::kParameter);
+  CreateProperties();
+}
+
+ShadowRealm::~ShadowRealm() {
+  while (HasCleanupHooks()) {
+    RunCleanup();
+  }
+}
+
+void ShadowRealm::MemoryInfo(MemoryTracker* tracker) const {
+  Realm::MemoryInfo(tracker);
+}
+
+v8::Local<v8::Context> ShadowRealm::context() const {
+  Local<Context> ctx = PersistentToLocal::Default(isolate_, context_);
+  DCHECK(!ctx.IsEmpty());
+  return ctx;
+}
+
+// V8 can not infer reference cycles between global persistent handles, e.g.
+// the Realm's Context handle and the per-realm function handles.
+// Attach the per-realm strong persistent values' lifetime to the context's
+// global object to avoid the strong global references to the per-realm objects
+// keep the context alive indefinitely.
+#define V(PropertyName, TypeName)                                              \
+  v8::Local<TypeName> ShadowRealm::PropertyName() const {                      \
+    return PersistentToLocal::Default(isolate_, PropertyName##_);              \
+  }                                                                            \
+  void ShadowRealm::set_##PropertyName(v8::Local<TypeName> value) {            \
+    HandleScope scope(isolate_);                                               \
+    PropertyName##_.Reset(isolate_, value);                                    \
+    v8::Local<v8::Context> ctx = context();                                    \
+    if (value.IsEmpty()) {                                                     \
+      ctx->Global()                                                            \
+          ->DeletePrivate(ctx, isolate_data()->per_realm_##PropertyName())     \
```

Comment:
I think we can just set the value to undefined here, which is less costly than delete.

Reply:
Done.